project1/server/agent.cpp

```diff
@@ -4,231 +4,415 @@
-      std::cerr << "Failed to go down a directory." << std::endl;
-      return ClientState::ERROR;
-    }
-  }
-
-  return SendResponse(kEmptyResponse) ? ClientState::ACTIVE
-                                      : ClientState::ERROR;
-}
-
-Agent::ClientState Agent::HandleRequest(
-    const ftp_messages::MakeDirRequest& request) {
-  std::cout << "Performing a MKDIR request." << std::endl;
-
-  // Create the directory.
-  if (!file_handler_->MakeDir(request.dir_name())) {
-    std::cerr << "Failed to create a directory." << std::endl;
-    return ClientState::ERROR;
-  }
-
-  return SendResponse(kEmptyResponse) ? ClientState::ACTIVE
-                                      : ClientState::ERROR;
-}
-
-Agent::ClientState Agent::HandleRequest(
-    const ftp_messages::PwdRequest& request) {
-  std::cout << "Performing a PWD request." << std::endl;
-
-  // Get the directory.
-  const auto& kCurrentDir = file_handler_->GetCurrentDir();
-
-  // Send the response.
-  Response response;
-  response.mutable_pwd()->set_dir_name(kCurrentDir);
-  return SendResponse(response) ? ClientState::ACTIVE : ClientState::ERROR;
-}
-
-Agent::ClientState Agent::HandleRequest(
-    const ftp_messages::QuitRequest& request) {
-  std::cout << "Got a QUIT request. Exiting." << std::endl;
-  // Indicate that we are finished with this client.
-  return ClientState::DISCONNECTED;
-}
+    using file_handler::ThreadSafeFileHandler;
+    using ftp_messages::Request;
+    using ftp_messages::Response;
+
+    namespace {
+
+        /// Generic empty response.
+        const Response kEmptyResponse{};
+
+    }  // namespace
+
+    Agent::Agent(int client_fd, std::unique_ptr<ThreadSafeFileHandler> file_handler,
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands)
+            : client_fd_(client_fd), active_commands_(active_commands), file_handler_(std::move(file_handler)) {}
+
+    Agent::Agent(int client_fd,
+                 std::shared_ptr<server_tasks::CommandIDs> active_commands) :
+            client_fd_(client_fd), active_commands_(active_commands) {}
+
+    Agent::~Agent() {
+        // Close the socket.
+        close(client_fd_);
+    }
+
+    bool Agent::Handle() {
+        ClientState client_state = ClientState::ACTIVE;
+
+
+        while (client_state == ClientState::ACTIVE) {
+            // Get the next message from the socket.
+            Request message;
+            client_state = ReadNextMessage(&message);
+
+            if (client_state == ClientState::ACTIVE) {
+                // Handle the message.
+                client_state = DispatchMessage(message);
+            }
+        }
+
+        return client_state != ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::ReadFileContents(ftp_messages::FileContents *fc, uint16_t command_id) {
+
+        while (!fc_parser_.HasCompleteMessage()) {
+            incoming_message_buffer_.resize(kClientBufferSize);
+
+            uint8_t buf[1000];
```

Comment:
You might not want to allocate such a large buffer on the stack. Note that my `incoming_mesage_buffer_` variable is a vector for a reason, since vectors allocate on the heap internally. It is also a class member so that we can re-use the same buffer, as long as it's sufficiently large (hence the call to `resize` above).

Also, I don't see much of a point in having two buffers, considering that `incoming_message_buffer_` was supposed to get overwritten each time around the loop anyway. (In the new version, you don't do that, which will completely break the message parsing.) Why not just continue reading directly into there?

To summarize, I think that the only changes necessary to this method are changing `kClientBufferSize` and adding the logic to check whether we should terminate every time around the loop.